src/main/java/io/jenkins/plugins/gitlabbranchsource/GitLabProjectSCMEvent.java

```diff
@@ -54,13 +66,65 @@ public String getSourceName() {
+        return navigator instanceof GitLabSCMNavigator && isMatch((GitLabSCMNavigator) navigator);
+    }
+
+    private boolean isMatch(@Nonnull GitLabSCMNavigator navigator) {
+        switch (getType()) {
+            case CREATED:
+                String projectPathWithNamespace = getPayload().getPathWithNamespace();
+                GitLabApi gitLabApi = null;
+                try {
+                    gitLabApi = GitLabHelper.apiBuilder(navigator.getServerName());
+                    GitLabOwner gitlabOwner = GitLabOwner.fetchOwner(gitLabApi, navigator.getProjectOwner());
+                    GitLabSCMNavigatorContext navigatorContext = new GitLabSCMNavigatorContext().withTraits(navigator.getTraits());
+                    List<Project> projects;
+                    // Cannot check the navigator projects because there could be empty projects/empty groups/new group or subgroups altogether
+                    if(gitlabOwner instanceof GitLabUser) {
+                        // Even returns the group projects owned by the user
+                        if(navigatorContext.wantSubgroupProjects()) {
+                            projects = gitLabApi.getProjectApi().getOwnedProjects();
+                        } else {
+                            projects = gitLabApi.getProjectApi().getUserProjects(navigator.getProjectOwner(), new ProjectFilter().withOwned(true));
+                        }
+                    } else {
+                        GroupProjectsFilter groupProjectsFilter = new GroupProjectsFilter();
+                        groupProjectsFilter.withIncludeSubGroups(navigatorContext.wantSubgroupProjects());
+                        // If projectOwner is a subgroup, it will only return projects in the subgroup
+                        projects = gitLabApi.getGroupApi().getProjects(navigator.getProjectOwner(), groupProjectsFilter);
+                    }
+                    for(Project project : projects) {
+                        if(project.getPathWithNamespace().equals(projectPathWithNamespace)) {
+                            return true;
+                        }
+                    }
+                    return false;
+                } catch (NoSuchFieldException | GitLabApiException e) {
+                    e.printStackTrace();
+                }
+                break;
+            case UPDATED:
+                if(navigator.getNavigatorProjects().contains(getPayload().getPathWithNamespace())) {
+                    // TODO: Not sure if this is the way to do it, need to check
+                    navigator.getNavigatorProjects().remove(getPayload().getPathWithNamespace());
+                    return true;
+                }
+                break;
+            case REMOVED:
+                if(navigator.getNavigatorProjects().contains(getPayload().getPathWithNamespace())) {
+                    navigator.getNavigatorProjects().remove(getPayload().getPathWithNamespace());
```

Comment:
NOOOOOOO

This is supposed to be a query method. It should not update anything at all. All this method should do is return `true` if the event is probably for the navigator (false positives are ok, false negatives are not)

Do not call out to remote systems... if you suspect it could be a match, return true and let the event listener confirm whether the event had an effect.
 
Do not change anything... you don't know what the event listener will want to do with the event